### PR TITLE
Saem specs update test organization

### PR DESCRIPTION
**Pull request by saem**
## Summary

Introduced a "bootstrap" specification that is meant to test a barebones
compiler implementation. This new arrangement of code sets the
revised tone of the specification.

## Details

This mostly rearranges the tests. While working on these particular
specs it was found that the identifier related lexer errors were
unclear. These have been revised and their generation is entirely driven
by the `lexer` module itself. `cli_reporter` now simply defers to it. As
part of the change `LexerReport` have lost a lot of their bloat.

Finally, removed testament's duplicate spec detection.
This only gets in the way of declaring a number of string
literal tests.

### CODE REVIEW

File: testament/specs.nim [hunk 1]
```diff
@@ -239,17 +239,12 @@ proc extractSpec(filename: string; spec: var TSpec): string =
...
       # `tests/stdlib/tbase64.nim` which contains BOM
       # (https://en.wikipedia.org/wiki/Byte_order_mark)
       const lineMax = 10
-      if a != -1:
-        raise newException(
-          ValueError,
-          "testament spec violation: duplicate `specStart` found: " &
-            $(filename, a, b, line))
-      elif line > lineMax:
+      if line > lineMax:
```
Comment (zerbina) [inline testament/specs.nim:247]: Could you include a mention of this change in behaviour in the PR message?

File: tests/lang/readme.md [hunk 2]
```diff
@@ -26,33 +24,26 @@ arguments, their behavior can be altered using keywords such as `break`,
...
-   way it does with `while`)
+   behavior within `while`, `block`, etc must be tested so we can ensure
+   they're consistent when used with iterators (equivalent coverage).
 
 When it comes to writing a test for the iterator itself we first start with
 the most basic iterator that does not accept any arguments, and can only
 `yield` values. We check if order of execution for statements is correct,
-and then `yield` does really return the same value. Iterator must also
-preserve it's state between execution transfers - we test it too.
+and then `yield` returns the correct value, and control is passed correctly,
```
Comment (zerbina) [inline tests/lang/readme.md]:
```suggestion
`yield` returns the correct value, and control is passed correctly,
```
Only a soft suggestion, but I think the sentence flows slightly better without.

File: tests/lang/readme.md [hunk 3]
```diff
@@ -1,23 +1,21 @@
...
-implementation of nim, or support a new backend. First you implement the
-most basic language features, then you progressively iterate them to
-correctly match their behavior with everything that has been added prior to
-that.
+specify the most fundamental language elements (comments, identifiers, blocks),
+then layers on additional capabilities building upon the previous. It is
+written as a conformance test suite starting from a barely functional barebones
+compiler up to a full featured current version.
+
+Generally speaking, each language constructure should be described in its own
```
Comment (zerbina) [inline tests/lang/readme.md]:
```suggestion
Generally speaking, each language construct should be described in its own
```

File: tests/lang/s00_bootstrap/readme.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Bootstrap
+
+This directory describes the minimal parts of the language if one was just
+starting on a barebone bootstrap compiler.
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/readme.md]:
```suggestion
starting on a barebones bootstrap compiler.
```

File: tests/lang/s00_bootstrap/t02_line_comments.nim [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# these are all comments
+
+# 20 + 30 # all sorts of things can be written here
+
+## this is also a comment it doesn't have to do anything different currntly and
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t02_line_comments.nim]:
```suggestion
## this is also a comment. It doesn't have to do anything different currently and
```

File: tests/lang/s00_bootstrap/t04_literal_strings.nim [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+"""
+
+## regular quoted strings can appear in multi-line strings because multi-line
+## variant look for `"""`
+discard """
+{
+  "key": "value",
+  3:     "three"
+}
+""" # it's not json, just a strong
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t04_literal_strings.nim]: I'm not entirely sure, but I think the last part of this comment is missing?

File: tests/lang/s00_bootstrap/t05_block_indentation.nim [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+"""
+
+block:      ## this is a block
+  discard   ## notice how we indented this line?
+
+## block: # alone, with nothing after would be an error
+## Nimskull is blankspace sensitive
+
+## indentation levels does not need to be consistent across the file, only
+## per level
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t05_block_indentation.nim]:
```suggestion
## indentation levels do not need to be consistent across the file, only
## per block
```
It's slightly clearer to use `block` here, I think.

File: tests/lang/s00_bootstrap/t06_block_label_identifiers.nim [hunk 8]
```diff
@@ -0,0 +1,24 @@
+discard """
+  description: "Describe block labels and valid label identifiers."
+
+  # valid label identifiers are the same rules for valid identifiers in
+  # general and we'll be using and buliding upon these rules going forward.
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t06_block_label_identifiers.nim]:
```suggestion
  # the valid identifier rules in general also apply to label identifiers and
  # we'll be using and building upon these rules going forward.
```
I think the intention is kept, but it flows a bit better.

File: tests/lang/s00_bootstrap/t06_block_label_identifiers.nim [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+"""
+
+## blocks can have labels
+block label:
+  discard
+
+## labels will later be used for control flow, but for now they're any valid
+## identifier, and we can use them as "titles" in the spec
+block whatever_valid_identifier_suits_our_fancy:
+  discard "consequtive underscores are `not__valid`"
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t06_block_label_identifiers.nim]:
```suggestion
  discard "consecutive underscores are `not__valid`"
```

File: tests/lang/s00_bootstrap/t08_builtins_echo.nim [hunk 10]
```diff
@@ -0,0 +1,12 @@
+discard """
+  description: "Specifies basic echo operation used for testing."
+  output: '''
+  Hello, World!
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t08_builtins_echo.nim]:
```suggestion
Hello, World!
```
This should fix the test failure.

File: tests/lang/s00_bootstrap/t08_builtins_echo.nim [hunk 11]
```diff
@@ -0,0 +1,12 @@
+discard """
+  description: "Specifies basic echo operation used for testing."
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t08_builtins_echo.nim]:
```suggestion
  description: "Specify basic echo operation used for testing."
```
For consistency with the descriptions of the previous tests.

File: tests/lang/s00_bootstrap/t10_builtins_doAssert_bool_literals_expressions.nim [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+"""
+
+block bool_literals:
+  ## again, discard is required for forwards compatibility
+  discard true
+  discard false
+
+block builtin_assert:
+  ## `doAssert` command is used throughout the specification to demonstrate
+  ## the correctness of an expression; ergo, it `doAssert` evaluates the
```
Comment (zerbina) [inline tests/lang/s00_bootstrap/t10_builtins_doAssert_bool_literals_expressions.nim]:
```suggestion
  ## the correctness of an expression; ergo, `doAssert` evaluates the
```

**Review by zerbina — APPROVED** [review]
Looks good to me, and I like the direction.

I left a few non-blocking suggestion for wording improvements and typo fixes.

File: testament/specs.nim (hunk 1 above)
Comment (saem) [reply testament/specs.nim:247]: Whoopsie, thanks and yes I can.

File: tests/lang/s00_bootstrap/t04_literal_strings.nim (hunk 6 above)
Comment (saem) [reply tests/lang/s00_bootstrap/t04_literal_strings.nim]: Meant to say, "just a string", instead of "strong". 😅

File: testament/specs.nim (hunk 1 above)
Comment (saem) [reply testament/specs.nim:247]: Done

**Pull request merged**